pkg/trace/context.go

```diff
@@ -0,0 +1,155 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package trace
+
+import (
+	"fmt"
+	"sync/atomic"
+	"time"
+
+	"github.com/Sirupsen/logrus"
+
+	"golang.org/x/net/context"
+)
+
+const OpTraceKey = "traceKey"
+
+// monotonic counter which inrements on Start()
+var opCount uint64
+
+type Operation struct {
+	context.Context
+	operation
+}
+
+type operation struct {
+	t     []tr
+	opNum uint64
+}
+
+func newOperation(ctx context.Context, opNum uint64, skip int, msg string) Operation {
+	op := operation{
+
+		// Can be used to trace based on this number which is unique per chain
+		// of operations
+		opNum: opNum,
+
+		// Start the trace.
+		t: []tr{*newTrace(msg, skip)},
+	}
+
+	// We need to be able to identify this operation across API (and process)
+	// boundaries.  So add the trace as a value to the embedded context.  We
+	// stash the values individually in the context because we can't assign
+	// the operation itself as a value to the embedded context (it's circular)
+	ctx = context.WithValue(ctx, OpTraceKey, op)
+
+	o := Operation{
+		Context:   ctx,
+		operation: op,
+	}
+
+	o.Debugf(o.t[0].beginHdr())
+	return o
+}
+
+// Creates a header string to be printed.
+func (o *Operation) header() string {
+	if Logger.Level >= logrus.DebugLevel {
+		return fmt.Sprintf("op=%d (delta:%s)", o.opNum, o.t[0].delta())
+	} else {
+		return fmt.Sprintf("op=%d", o.opNum)
+	}
+}
+
+// Err returns a non-nil error value after Done is closed.  Err returns
+// Canceled if the context was canceled or DeadlineExceeded if the
+// context's deadline passed.  No other values for Err are defined.
+// After Done is closed, successive calls to Err return the same value.
+func (o Operation) Err() error {
+
+	// Walk up the contexts from which this context was created and get their errors
+	if err := o.Context.Err(); err != nil {
+		// Print the error
+		o.Errorf("%s: %s error: %s", o.t[0].endHdr(), o.t[0].msg, err)
+
+		// Walk the stack and end with where this was called from
+		for _, t := range append(o.t, *newTrace("Err", 2)) {
+			Logger.Errorf("\t%s:%d %s", t.funcName, t.lineNo, t.msg)
+		}
+
+		return err
+	}
+
+	return nil
+}
+
+func (o *Operation) Infof(format string, args ...interface{}) {
+	Logger.Infof("%s: %s", o.header(), fmt.Sprintf(format, args...))
+}
+
+func (o *Operation) Debugf(format string, args ...interface{}) {
+	Logger.Debugf("%s: %s", o.header(), fmt.Sprintf(format, args...))
+}
+
+func (o *Operation) Errorf(format string, args ...interface{}) {
+	Logger.Errorf("%s: %s", o.header(), fmt.Sprintf(format, args...))
+}
+
+func (o *Operation) newChild(ctx context.Context, msg string) Operation {
+	child := newOperation(ctx, o.opNum, 4, msg)
+	t := child.t[0]
+	child.t = append(o.t, t)
+	return child
+}
+
+// Add tracing info to the context.
+func NewOperation(ctx context.Context, msg string) Operation {
+	return newOperation(ctx, atomic.AddUint64(&opCount, 1), 3, msg)
```

Comment:
please add the pid here as well so we don't get mixed opid's across log files/processes.